unidler.py

```diff
@@ -48,14 +83,126 @@ def respond(self, status, body=None, headers={}):
 
 
-class UnidlerServer(ThreadingMixIn, HTTPServer):
+class DeploymentNotFound(Exception):
     pass
 
 
-def run(host='0.0.0.0', port=8080):
-    unidler = UnidlerServer((host, port), RequestHandler)
-    print(f'Unidler listening on {host}:{port}')
-    unidler.serve_forever()
+class IngressNotFound(Exception):
+    pass
+
+
+def unidle_deployment(hostname):
+    ingresses = build_ingress_lookup()
+
+    with ingress_for_host(hostname, ingresses) as ing:
+        with deployment_for_ingress(ing) as deployment:
+            restore_replicas(deployment)
+            unmark_idled(deployment)
+            enable_ingress(ing)
+
+    with ingress(UNIDLER, 'default', ingresses) as unidler_ingress:
+        remove_host_rule(hostname, unidler_ingress)
+
+
+def build_ingress_lookup():
+    ingresses = client.ExtensionsV1beta1Api().list_ingress_for_all_namespaces()
+    return dict(
+        ((ingress.metadata.name, ingress.metadata.namespace), ingress)
+        for ingress in ingresses.items)
+
+
+@contextlib.contextmanager
+def ingress_for_host(hostname, ingresses):
+    for ingress in ingresses.values():
+        name = ingress.metadata.name
+        namespace = ingress.metadata.namespace
+
+        if (name, namespace) == (UNIDLER, 'default'):
+            continue
+
+        for rule in ingress.spec.rules:
+            if rule.host == hostname:
+                log.debug(
+                    f'Found ingress for {hostname}: {name} '
+                    f'in namespace {namespace}')
+                yield ingress
+                return write_ingress_changes(ingress)
+
+    raise IngressNotFound(f'Ingress for host {hostname} not found')
+
+
+def write_ingress_changes(ingress):
+    log.debug(
+        f'Writing changes to ingress {ingress.metadata.name} '
+        f'in namespace {ingress.metadata.namespace}')
+    client.ExtensionsV1beta1Api().patch_namespaced_ingress(
+        ingress.metadata.name,
+        ingress.metadata.namespace,
+        ingress)
+
+
+@contextlib.contextmanager
+def deployment_for_ingress(ingress):
+    name = ingress.metadata.name
+    namespace = ingress.metadata.namespace
+
+    try:
+        deployment = client.AppsV1beta1Api().read_namespaced_deployment(
+            name,
+            namespace)
+        log.debug(f'Found deployment {name} in namespace {namespace}')
+
+        yield deployment
+
+        write_changes(deployment)
+
+    except kubernetes.client.rest.ApiException as error:
+        raise DeploymentNotFound(f'Deployment {name} not found in {namespace}')
+
+
+def write_changes(deployment):
+    log.debug(
+        f'Writing changes to deployment {deployment.metadata.name} '
+        f'in namespace {deployment.metadata.namespace}')
+    client.AppsV1beta1Api().replace_namespaced_deployment(
+        deployment.metadata.name,
+        deployment.metadata.namespace,
+        deployment)
+
+
+def restore_replicas(deployment):
+    idled_at, replicas = deployment.metadata.annotations[IDLED_AT].split(',')
```

Comment:
I'm not sure if use this annotation to keep the time when the deployment was idled AND the number of replicas before being idled is the best. Maybe it's too late/too annoying to change as well. Maybe it's not worth having another annotation just to keep this counter.

Ignore me, just thinking aloud really.

Reply:
You're right, there's no particular reason to use a single annotation, could have two. In fact, the timestamp is unused at the moment - the plan was to use it to prevent idling a deployment that is currently being idled, but that may not be necessary.